### PR TITLE
fix(setting): add a default setting `backupExecutionTimeout`

### PR DESCRIPTION
**Pull request by mantissahz**
#### Which issue(s) this PR fixes:

Issue # longhorn/longhorn#8954 longhorn/longhorn#8319

### CODE REVIEW

File: chart/README.md [hunk 1]
```diff
@@ -284,6 +284,7 @@ During installation, you can either allow Longhorn to use the default system set
 | defaultSettings.backingImageRecoveryWaitInterval | Number of seconds that Longhorn waits before downloading a backing image file again when the status of all image disk files changes to "failed" or "unknown". |
 | defaultSettings.backupCompressionMethod | Setting that allows you to specify a backup compression method. |
 | defaultSettings.backupConcurrentLimit | Maximum number of worker threads that can concurrently run for each backup. |
+| defaultSettings.backupExecuteTimeout | Number of minutes that Longhorn sets the timeout for the backup engine execution. When the value is "0", automatic deletion is disabled. |
```
Comment (derekbit) [inline chart/README.md:287]: How about `backupExecutionTimeout`?
What does `automatic deletion is disabled` mean?

Comment (mantissahz) [reply chart/README.md]: Renamed.

>What does automatic deletion is disabled mean?

Updated for disabling the timeout.

File: chart/questions.yaml [hunk 2]
```diff
@@ -374,6 +374,13 @@ questions:
     type: int
     min: 0
     default: 1440
+  - variable: defaultSettings.backupExecutionTimeout
+    label: Backup Execute Timeout
+    description: "Number of minutes that Longhorn sets the timeout for the backup engine execution. When the value is \"0\", disable the timeout."
```
Comment (derekbit) [inline chart/questions.yaml:379]: What does `disable the timeout` mean?

Comment (mantissahz) [reply chart/questions.yaml:379]: The request/operation (binary command) will not be interrupted by a timeout (there is no timeout).

Comment (derekbit) [reply chart/questions.yaml:379]: Got it. Can we make it clearer?
For example, `If timeout is -1, it will wait infinitely until .....`.

Comment (mantissahz) [reply chart/questions.yaml]: Modified. Thanks

File: chart/questions.yaml [hunk 3]
```diff
@@ -374,6 +374,13 @@ questions:
     type: int
     min: 0
     default: 1440
+  - variable: defaultSettings.backupExecutionTimeout
+    label: Backup Execute Timeout
+    description: "Number of minutes that Longhorn sets the timeout for the backup engine execution. If timeout is set to 0, it will wait indefinitely until the execution returns."
```
Comment (derekbit) [inline chart/questions.yaml:379]: I'm thinking whether 0 is reasonable. WDYT? @c3y1huang @ChanYiLin @shuo-wu

Comment (shuo-wu) [reply chart/questions.yaml:379]: Although backup creation is an async call, I don't think an infinite wait is the best default setting... But on the other hand, do we support interrupting the progress when the timeout reaches?

Comment (mantissahz) [reply chart/questions.yaml:379]: >I don't think an infinite wait is the best default setting

@shuo-wu, sorry for misleading you, the default setting is 1 minute.

>But on the other hand, do we support interrupting the progress when the timeout reaches?

I think we only return an error when the timeout reaches.

Comment (shuo-wu) [reply chart/questions.yaml:379]: > > But on the other hand, do we support interrupting the progress when the timeout reaches?
>
> I think we only return an error when the timeout reaches.

I mean, how can Longhorn stop it if the backup creation actually starts after the timeout?

Comment (mantissahz) [reply chart/questions.yaml:379]: If the backup creation actually starts after the timeout (the Error `DeadlineExceeded`), the backup state will be marked as `pending` and wait for the next reconciling.

https://github.com/longhorn/longhorn-manager/blob/abb8f260796dcc5ee1ab84c5a20385217aeb7243/engineapi/backup_monitor.go#L77-L119

Longhorn would not stop the backup creation when the timeout reaches.

But the setting `backupExecutionTimeout` only handles the backup binary operations like `engines/longhorn backup ls --volume-only`, `backup inspect-volume`, and so on.

Comment (mantissahz) [reply chart/questions.yaml:379]: And I think disabling the timeout here is not a good idea and we have [ExecuteEngineBinaryWithoutTimeout](https://github.com/longhorn/longhorn-manager/blob/abb8f260796dcc5ee1ab84c5a20385217aeb7243/engineapi/backups.go#L144) already if the operations need to wait until they complete.

Comment (mantissahz) [reply chart/questions.yaml:379]: Yes, WDYT?
Some operations need a timeout.

Comment (derekbit) [reply chart/questions.yaml:379]: Sounds good to me.

Comment (mantissahz) [reply chart/questions.yaml]: Modified the default value to 1, and the value can not be smaller than 1.

File: chart/questions.yaml [hunk 4]
```diff
@@ -374,6 +374,13 @@ questions:
     type: int
     min: 0
     default: 1440
+  - variable: defaultSettings.backupExecutionTimeout
+    label: Backup Execute Timeout
+    description: "Number of minutes that Longhorn sets the timeout for the backup engine execution. The default value is \"1\"."
+    group: "Longhorn Default Settings"
+    type: int
+    min: 0
```
Comment (derekbit) [inline chart/questions.yaml:382]: Does it mean `0` is allowed?

Comment (derekbit) [reply chart/questions.yaml:382]: In addition, why not use seconds?

Comment (mantissahz) [reply chart/questions.yaml]: >Does it mean 0 is allowed?

Yes, modified to 1

>In addition, why not use seconds?

According to the user's response, they usually set a long timeout (10 minutes).
Either one (second or minute) is fine with me. WDYT?

Comment (derekbit) [reply chart/questions.yaml]: > And the origin time unit from the library go-common-libs is minute.

Can you guide me the link? Thank you.

Comment (mantissahz) [reply chart/questions.yaml]: here it is https://github.com/longhorn/go-common-libs/blob/907f57efd48f43e1ca32f4ffa634d3c6b55b5a54/types/exec.go#L14

Comment (derekbit) [reply chart/questions.yaml]: Thank you.

File: chart/questions.yaml [hunk 5]
```diff
@@ -374,6 +374,13 @@ questions:
     type: int
     min: 0
     default: 1440
+  - variable: defaultSettings.backupExecutionTimeout
+    label: Backup Execute Timeout
+    description: "Number of minutes that Longhorn sets the timeout for the backup engine execution. The default value is \"1\"."
```
Comment (derekbit) [inline chart/questions.yaml:379]: @jillian-maroket @jhkrug Can you review the wording? Thank you.

File: chart/README.md [hunk 6]
```diff
@@ -325,6 +325,7 @@ During installation, you can either allow Longhorn to use the default system set
 | defaultSettings.replicaZoneSoftAntiAffinity | Setting that allows Longhorn to schedule new replicas of a volume to nodes in the same zone as existing healthy replicas. Nodes that do not belong to any zone are treated as existing in the zone that contains healthy replicas. When identifying zones, Longhorn relies on the label "topology.kubernetes.io/zone=<Zone name of the node>" in the Kubernetes node object. |
 | defaultSettings.restoreConcurrentLimit | Maximum number of worker threads that can concurrently run for each restore operation. |
 | defaultSettings.restoreVolumeRecurringJobs | Setting that restores recurring jobs from a backup volume on a backup target and creates recurring jobs if none exist during backup restoration. |
+| defaultSettings.rwxVolumeFastFailover | Turn on logic to detect and move RWX volumes quickly on node failure. |
```
Comment (jillian-maroket) [inline chart/README.md:328]: The first part is unclear so here are my best guesses:

Option 1: Setting that allows Longhorn to detect node failure and immediately migrate affected RWX volumes.
Option 2: Setting that allows Longhorn to quickly migrate RWX volumes affected by node failure.

File: chart/questions.yaml (hunk 5 above)
Comment (jillian-maroket) [reply chart/questions.yaml:379]: I did not find any mentions of a "backup engine" in the doc. The draft is unclear so here are my best guesses:

Option 1: Number of minutes that Longhorn allows for backup creation. The default value is \"1\"
Option 2: Number of minutes the backup engine is allowed to run. The default value is \"1\".

**Review by jillian-maroket — COMMENTED** [review]
I provided a few options instead of definite suggestions because the drafts are unclear.

File: chart/questions.yaml (hunk 5 above)
Comment (derekbit) [reply chart/questions.yaml:379]: @mantissahz Should be `backup execution`?

Comment (mantissahz) [reply chart/questions.yaml]: >@mantissahz Should be backup execution?

Yes, and modified like `Option 1`
```txt
Number of minutes that Longhorn allows for backup execution. The default value is "1"
```

File: chart/README.md (hunk 6 above)
Comment (mantissahz) [reply chart/README.md]: Modified to `Option 1`. Thank you.

**Review by jillian-maroket — APPROVED** [review]

**Review by derekbit — APPROVED** [review]
LGTM

**Comment by mantissahz** [comment]
@Mergifyio backport v1.7.x

**Pull request merged**